scripts/deploy.sh

```diff
@@ -7,17 +7,88 @@ pull_latest() {
+  fi
+  mkdir -p $build_folder
+
+  # We want latest commit hash, so we do a shallow copy of the repositories (and not simply a wget)
+  git clone --depth=1 https://github.com/cubing/wca-regulations.git $build_folder/wca-regulations
+  git clone --depth=1 https://github.com/cubing/wca-regulations-translations.git $build_folder/wca-regulations-translations
+  git_reg_hash=`git -C $build_folder/wca-regulations rev-parse --short HEAD`
+  git_translations_hash=`git -C $build_folder/wca-regulations-translations rev-parse --short HEAD`
+
+  rebuild_regulations=1
+  rebuild_translations=1
+  # Check if the cloned regulations match the current version
+  if [ -r $regs_version ] && [ "`cat $regs_version`" == "$git_reg_hash" ]; then
+    rebuild_regulations=0
+  fi
+  # Check if the cloned translations match the current version
+  if [ -r $translations_version ] && [ "`cat $translations_version`" == "$git_translations_hash" ]; then
+    rebuild_translations=0
+  fi
+  if [ $rebuild_regulations -eq 0 ] && [ $rebuild_translations -eq 0 ]; then
+    echo "WCA Regulations and translations are up to date."
+    return
+  fi
+
+  # Else we have to rebuild something
+
+  # This save tracked files that may have unstashed changes too
+  cp -r $regs_folder $build_folder
+
+  # Checkout data (scramble programs, history)
+  git checkout origin/regulations-data $build_folder
+  git reset HEAD $build_folder
+
+  languages=`wrc-languages`
+  if [ $rebuild_translations -eq 1 ]; then
+    # Clean up translations directories
+    find $build_folder/regulations/translations ! -name 'translations' -type d -exec rm -rf {} +
+    # Rebuild all translations
+    for kind in html pdf; do
+      for l in $languages; do
+        inputdir=$build_folder/wca-regulations-translations/${l}
+        outputdir=$build_folder/regulations/translations/${l}
+        mkdir -p $outputdir
+        echo "Doing "${kind}" for language "${l}
+        wrc --target=$kind -l $l -o $outputdir -g $git_translations_hash $inputdir
+      done
+    done
+    # Update version built
+    echo $git_translations_hash > $build_folder/regulations/translations/version
+  else
+    echo "Translations are up to date."
+  fi
+
+  outputdir=$build_folder/regulations
+  if [ $rebuild_regulations -eq 1 ]; then
+    # Clean up regulations directory files
+    find $build_folder/regulations -maxdepth 1 -type f -exec rm -f {} +
+    # Rebuild Regulations
+    wrc --target=json -o $outputdir -g $git_reg_hash $build_folder/wca-regulations
+    wrc --target=html -o $outputdir -g $git_reg_hash $build_folder/wca-regulations
+    wrc --target=pdf -o $outputdir -g $git_reg_hash $build_folder/wca-regulations
+    # Update version built
+    echo $git_reg_hash > $build_folder/regulations/version
+  else
+    echo "Regulations are up to date"
+  fi
+
   if [ -a $regs_folder ]; then
-    rm -rf $regs_folder-todelete
-    mv $regs_folder $regs_folder-todelete
+    # Remove previous build
+    rm -rf $regs_folder
```

Comment:
The convoluted order of stuff in the old file was actually pretty intentional. We have `$outputdir` and `$regs_folder`. We want `$outputdir` to end up in place of `$regs_folder`. There are a few ways to achieve this:

## Simplest (what you did here)
1. Delete `$regs_folder`.
2. Move `$outputdir` to `$regs_folder`.

However, deleting files actually can take some time (I believe `rm -r` has to recurse through the whole directory and `unlink` a bunch of files. During this time, the regulations are in a broken state. So, I wrote things in this slightly more convoluted order:

## More convoluted
1. Move `$regs_folder` to some temp name that you plan to delete later.
2. Move `outputdir` to `$regs_folder`.
3. Now we finally delete the old `$regs_folder`.

It is my understanding that both steps 1) and 2) above should be quick, atomic operations. Then we can spend all day `unlink`-ing files in 3), but at least the website won't be down.

I may be very wrong about some of the stuff here. This also may be a classic example of premature optimization. Let me know what you think.